File: 백준/Gold/11265. 끝나지 않는 파티/끝나지 않는 파티.py
from sys import stdin
import heapq

n, m = map(int, stdin.readline().split())
parties = [[]] + [[0]+list(map(int, stdin.readline().split())) for _ in range(n)]
INF = int(1e9)
costs = [[INF] * (n + 1) for _ in range(n + 1)]

def shortest_path(sour: int) -> None:
    global costs, parties, n
    heap = [[0, sour]]
    costs[sour][sour] = 0

    while heap:
        cnt, cur = heapq.heappop(heap)
        if cnt > costs[sour][cur]:
            continue
        for des in range(1, n + 1):
            dist = parties[cur][des]
            next_value = dist + cnt
            if dist and costs[sour][des] > next_value:
                costs[sour][des] = next_value
                heapq.heappush(heap, [next_value, des])

for i in range(1, n + 1):
    shortest_path(i)

for _ in range(m):
    sour, des, interval = map(int, stdin.readline().split())
    print('Enjoy other party' \
          if costs[sour][des] <= interval and interval != INF else 'Stay here')

File: 백준/Gold/11265. 끝나지 않는 파티/끝나지 않는 파티.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

public class Main {
	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		String[] size = br.readLine().split(" ");
		int n = Integer.parseInt(size[0]);
		int m = Integer.parseInt(size[1]);
		int[][] graph = new int[n+1][n+1];
		
		for(int i=1 ; i<=n ; i++) {
			String[] values = br.readLine().split(" ");
			for(int j=1 ; j<=n ; j++) {
				graph[i][j] = Integer.parseInt(values[j-1]);
			}
		}
		
		for(int k=1 ; k<=n ; k++) {
			for(int i=1 ; i<=n ; i++) {
				for(int j=1 ; j<=n ; j++) {
					graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]);
				}
			}
		}
		
		while(m-- > 0) {
			String[] str = br.readLine().split(" ");
			int sour = Integer.parseInt(str[0]), des = Integer.parseInt(str[1]), interval = Integer.parseInt(str[2]);
			bw.write(graph[sour][des] <= interval ? "Enjoy other party\n" : "Stay here\n");
		}
		
		bw.flush();
		br.close();
		bw.close();
	}

}
